OpenNefia.Content/UI/Hud/HudMessageWindow.cs

```diff
@@ -0,0 +1,243 @@
+﻿using OpenNefia.Content.UI.Element;
+using OpenNefia.Content.UI.Element.Containers;
+using OpenNefia.Core.Locale;
+using OpenNefia.Core.Maths;
+using OpenNefia.Core.Rendering;
+using OpenNefia.Core.UI.Element;
+using OpenNefia.Core.Utility;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OpenNefia.Content.UI.Hud
+{
+    public abstract class MessageTag
+    {
+    }
+
+    public class ColorMessageTag : MessageTag
+    {
+        public Color Color;
+
+        public ColorMessageTag(Color? color)
+        {
+            Color = color ?? UiColors.MesWhite;
+        }
+    }
+
+    public class TextMessageTag : MessageTag
+    {
+        public string Message;
+
+        public TextMessageTag(string message)
+        {
+            Message = message;
+        }
+    }
+
+    public class NewlineMessageTag : MessageTag
+    {
+
+    }
+
+    public class IconMessageTag : MessageTag
+    {
+
+    }
+
+    public class FormattedMessage
+    {
+        public List<MessageTag> Tags;
+        public FormattedMessage(IEnumerable<MessageTag>? tags = null)
+        {
+            Tags = tags?.ToList() ?? new();
+        }
+        public void AddTag(MessageTag tag)
+        {
+            Tags.Add(tag);
+        }
+        public static FormattedMessage WithColor(string text, Color? color = null)
+        {
+            var mes = new FormattedMessage();
+            mes.AddTag(new ColorMessageTag(color));
+            mes.AddTag(new TextMessageTag(text));
+            return mes;
+        }
+    }
+
+    public class HudMessageWindow : UiElement, IHudMessageWindow
+    {
+        public class MessageContainer : UiHorizontalContainer
+        {
+            public bool HasContent => Entries.Any(x => x.Element is MessageText);
+            public void SetOpacities(byte opacity)
+            {
+                foreach(var item in Entries)
+                {
+                    if (item.Element is MessageText text && text.ChangeOpacity)
+                        text.Color = text.Color.WithAlpha(opacity);
+                }
+            }
+        }
+
+        public class MessageText : UiText
+        {
+            public bool ChangeOpacity { get; set; }
+
+            public MessageText(string text, bool changeOpacity = true) : base(UiFonts.MessageText, text)
+            {
+                ChangeOpacity = changeOpacity;
+            }
+        }
+
+        private CircularBuffer<FormattedMessage> Messages;
+        private UiContainer BacklogContainer;
+        private UiContainer MessageBoxContainer;
+
+        private const int BacklogLines = 26;
+        private const int MessageBoxLines = 4;
+
+        private readonly byte[] MessageOpacities =
+        {
+            130,
```

Comment:
Small note: these can be configured in the options in vanilla. The option is a single integer "step" to decrement the opacity of each successive message line by.

Reply:
changed in 9eb449f
changed the calculation a little, with the simple decrement the opacities weren't lining up with vanilla
it seems closer with what I have now